bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/builder/DataLayerSupplierImpl.java

```diff
@@ -0,0 +1,491 @@
+     */
+    @Nullable
+    private Supplier<String[]> shownItemsSupplier;
+
+    /**
+     * The asset data value supplier.
+     */
+    @Nullable
+    private Supplier<AssetData> assetDataSupplier;
+
+    /**
+     * The URL value supplier.
+     */
+    @Nullable
+    private Supplier<String> urlSupplier;
+
+    /**
+     * The format value supplier.
+     */
+    @Nullable
+    private Supplier<String> formatSupplier;
+
+    /**
+     * The tags value supplier.
+     */
+    @Nullable
+    private Supplier<String[]> tagsSupplier;
+
+    /**
+     * The template path value supplier.
+     */
+    @Nullable
+    private Supplier<String> templatePathSupplier;
+
+    /**
+     * The language value supplier.
+     */
+    @Nullable
+    private Supplier<String> languageSupplier;
+
+    /**
+     * Construct a wrapper for a {@link DataLayerSupplier}.
+     *
+     * @param dataLayerSupplier The data layer supply to wrap.
+     */
+    public DataLayerSupplierImpl(@NotNull final DataLayerSupplier dataLayerSupplier) {
+        this.wrappedSupplier = dataLayerSupplier;
+    }
+
+    /**
+     * Construct a DataLayerSupplier from existing ComponentData.
+     *
+     * @param data The existing component data.
+     * @return A new DataLayerSupplier that uses the field values from the provided component data.
+     */
+    public static DataLayerSupplier extend(@NotNull final ComponentData data) {
+        // set component field value suppliers
+        final DataLayerSupplierImpl supplier = new DataLayerSupplierImpl(DataLayerSupplier.EMPTY_SUPPLIER)
+            .setId(data::getId)
+            .setType(data::getType)
+            .setLastModifiedDate(data::getLastModifiedDate)
+            .setParentId(data::getParentId)
+            .setTitle(data::getTitle)
+            .setDescription(data::getDescription)
+            .setText(data::getText)
+            .setLinkUrl(data::getLinkUrl);
+
+        // set page field value suppliers
+        if (data instanceof PageData) {
+            PageData pageData = (PageData) data;
+            supplier.setUrl(pageData::getUrl)
+                .setTags(pageData::getTags)
+                .setTemplatePath(pageData::getTemplatePath)
+                .setLanguage(pageData::getLanguage);
+        }
+
+        // set container field value suppliers
+        if (data instanceof ContainerData) {
+            supplier.setShownItems(((ContainerData) data)::getShownItems);
+        }
+
+        // set image component field value suppliers
+        if (data instanceof ImageData) {
+            supplier.setAssetData(((ImageData) data)::getAssetData);
+        }
+
+        return supplier;
+    }
+
+    /**
+     * Construct a DataLayerSupplier from existing AssetData.
+     *
+     * @param data The existing asset data.
+     * @return A new DataLayerSupplier that uses the field values from the provided asset data.
+     */
+    public static DataLayerSupplier extend(@NotNull final AssetData data) {
```

Comment:
Should we also introduce an `extend()` method for container, page and image components?

Reply:
AssetData is the odd one out, it doesn't extend ComponentData (not being a component).
The previous method `public static DataLayerSupplier extend(@NotNull final ComponentData data)` is used for all other component data.

There is functionality for extending component data using `DataLayerBuilder.extending(componentData.asX()` (where X is the data type `Image`, `Page`, `Container`, etc. 

One issue is that the component models do not tell you what kind of data is going to be returned. I.e. the `Image` component model does not indicate that it will return `ImageData`. Therefore, if a custom model is created which extends the `Image` model via sling delegation pattern, it is impossible for the builder to know that the data from the extended model is `ImageData`. Thus the implementer is required to specify the type, e.g. 
```JAVA
DataLayerBuilder.extending(parentModel.getComponentData()).asImageComponent()
```

This is also the reason why the `instanceof` checks are needed in the `extend(ComponentData)` method. 
I really hate `instanceof` checks here (and pretty much everywhere) - but if the different types of `ComponentData` were separated back into their own implementation classes then someone extending a core component model would need to make a cast, and that cast wouldn't be safe. 

To resolve this issue in a way that doesn't feel like a hack, the component model interfaces should be explicit about the type of data they are going to return; however, that is a breaking change. 

This rational also covers previous comments as to why the different data types are all lumped into one class (`ComponentDataImpl`). Also, while reviewing this, please keep in mind that `ComponentDataImpl` is not exposed - outside of this package you can only interact with the `DataLayerBuilder`, which will only present you with the methods applicable for the specified data type.

Reply:
ok, thanks for your explanations, it makes sense to keep it as it is.